test/infrastructure/vcsim/controllers/envvar_controller.go

```diff
@@ -287,14 +292,68 @@ func clusterEnvVarSpecGovmomiVariables(c *vcsimv1.ClusterEnvVarSpec) map[string]
 	}
 
 	// NOTE: omitting cluster Name intentionally because E2E tests provide this value in other ways
-	vars["VSPHERE_DATACENTER"] = vcsimhelpers.DatacenterName(datacenter)
+	// The var VSPHERE_COMPUTE_CLUSTER cannot be used as moid during the tests.
 	vars["VSPHERE_COMPUTE_CLUSTER"] = vcsimhelpers.ClusterName(datacenter, cluster)
-	vars["VSPHERE_DATASTORE"] = vcsimhelpers.DatastoreName(datastore)
-	vars["VSPHERE_FOLDER"] = vcsimhelpers.VMFolderName(datacenter)
-	vars["VSPHERE_NETWORK"] = vcsimhelpers.NetworkPath(datacenter, vcsimhelpers.DefaultNetworkName)
-	vars["VSPHERE_RESOURCE_POOL"] = vcsimhelpers.ResourcePoolPath(datacenter, cluster)
-	vars["VSPHERE_TEMPLATE"] = vcsimhelpers.VMPath(datacenter, template)
-	return vars
+
+	if !moid {
+		vars["VSPHERE_DATACENTER"] = vcsimhelpers.DatacenterName(datacenter)
+		vars["VSPHERE_DATASTORE"] = vcsimhelpers.DatastoreName(datastore)
+		vars["VSPHERE_FOLDER"] = vcsimhelpers.VMFolderName(datacenter)
+		vars["VSPHERE_RESOURCE_POOL"] = vcsimhelpers.ResourcePoolPath(datacenter, cluster)
+		vars["VSPHERE_NETWORK"] = vcsimhelpers.NetworkPath(datacenter, vcsimhelpers.DefaultNetworkName)
+		vars["VSPHERE_TEMPLATE"] = vcsimhelpers.VMPath(datacenter, template)
+		return vars, nil
+	}
+
+	params := session.NewParams().
+		WithServer(existing["VSPHERE_SERVER"]).
+		WithThumbprint(existing["VSPHERE_TLS_THUMBPRINT"]).
+		WithUserInfo(existing["VSPHERE_USERNAME"], existing["VSPHERE_PASSWORD"])
+
+	s, err := session.GetOrCreate(ctx, params)
+	if err != nil {
+		return vars, errors.Errorf("error creating test session: %v", err)
+	}
+
+	dc, err := s.Finder.Datacenter(ctx, vcsimhelpers.DatacenterName(datacenter))
+	if err != nil {
+		return vars, errors.Errorf("failed to locate datacenter reference: %v", err)
+	}
+
+	vars["VSPHERE_DATACENTER"] = dc.Reference().String()
```

Comment:
copying @dougm comment here:
The changed API doc says "name, inventory path or the MOID". But Reference().String() here (and so on below) is the MOR (Managed Object Reference). And Reference().Value is the MOID. The govmomi Finder changes support both, see: https://github.com/vmware/govmomi/blob/a2c22f0e33da798414846a99fd2f55522a405b13/object/common.go#L160-L166